File: backend/app/api/v1/endpoints/pos.py
from typing import List, Optional
from datetime import date
from uuid import UUID
from fastapi import APIRouter, Depends, HTTPException, Query, Path, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import (
    get_current_user,
    UserScopedDbDep,
    AuthenticatedDbDep
)
from app.services.pos_service import POSService
from app.schemas.pos import (
    # Transactions
    CreateTransactionRequest,
    TransactionResponse,
    TransactionUpdate,
    ProcessPaymentRequest,
    PaymentResponse,
    QRPaymentInitResponse,
    PaymentStatusResponse,
    VoidTransactionRequest,
    SearchTransactionRequest,
    BatchTransactionRequest,
    BatchTransactionResponse,
    
    # Shifts
    OpenShiftRequest,
    ShiftResponse,
    CloseShiftRequest,
    ShiftSummaryResponse,
    
    # Receipts
    PrintReceiptRequest,
    ReceiptResponse,
    ReceiptTemplateCreate,
    ReceiptTemplateUpdate,
    ReceiptTemplateResponse,
    
    # Categories & Modifiers
    POSCategoryCreate,
    POSCategoryUpdate,
    POSCategoryResponse,
    ProductModifierCreate,
    ProductModifierUpdate,
    ProductModifierResponse,
    
    # Reports
    DailySummaryRequest,
    DailySummaryResponse,
    ProductSalesReport,
    
    # Others
    QuickProductResponse,
    PrinterStatusResponse,
    PrintTestRequest
)

router = APIRouter()

# ============= Transaction Endpoints =============

@router.post("/transactions/create", response_model=TransactionResponse)
async def create_transaction(
    request: CreateTransactionRequest,
    db: AuthenticatedDbDep,
    current_user: dict = Depends(get_current_user)
):
    """Create new POS transaction"""
    service = POSService(db)
    return await service.create_transaction(request, current_user['id'])

@router.get("/transactions/{transaction_id}", response_model=TransactionResponse)
async def get_transaction(
    db: AuthenticatedDbDep,
    transaction_id: UUID = Path(...),
    current_user: dict = Depends(get_current_user)
):
    """Get transaction details"""
    service = POSService(db)
    return await service._get_transaction_with_items(transaction_id)

@router.put("/transactions/{transaction_id}", response_model=TransactionResponse)
async def update_transaction(
    db: AuthenticatedDbDep,
    transaction_id: UUID = Path(...),
    request: TransactionUpdate = ...,
    current_user: dict = Depends(get_current_user)
):
    """Update transaction details"""
    service = POSService(db)
    # TODO: Implement update method
    raise HTTPException(status_code=501, detail="Not implemented")

@router.post("/transactions/{transaction_id}/payment")
async def process_payment(
    db: AuthenticatedDbDep,
    transaction_id: UUID = Path(...),
    request: ProcessPaymentRequest = ...,
    current_user: dict = Depends(get_current_user)
):
    """Process payment for transaction"""
    service = POSService(db)
    result = await service.process_payment(transaction_id, request, current_user['id'])
    return result

@router.get("/transactions/{transaction_id}/payment-status", response_model=PaymentStatusResponse)
async def check_payment_status(
    db: AuthenticatedDbDep,
    transaction_id: UUID = Path(...),
    current_user: dict = Depends(get_current_user)
):
    """Check QR payment status (polling endpoint)"""
    # For now, return a simple pending status
    # TODO: Implement proper payment status checking
    return PaymentStatusResponse(
        transaction_id=transaction_id,
        payment_status=PaymentStatus.PENDING,
        payment_received=None,
        action=None
    )

@router.post("/transactions/{transaction_id}/void", response_model=TransactionResponse)
async def void_transaction(
    db: AuthenticatedDbDep,
    transaction_id: UUID = Path(...),
    request: VoidTransactionRequest = ...,
    current_user: dict = Depends(get_current_user)
):
    """Void a transaction"""
    service = POSService(db)
    return await service.void_transaction(transaction_id, request, current_user['id'])

@router.post("/transactions/{transaction_id}/print-receipt", response_model=ReceiptResponse)
async def print_receipt(
    db: AuthenticatedDbDep,
    transaction_id: UUID = Path(...),
    request: PrintReceiptRequest = ...,
    current_user: dict = Depends(get_current_user)
):
    """Print or reprint receipt"""
    service = POSService(db)
    return await service.print_receipt(transaction_id, request)

@router.post("/transactions/search", response_model=List[TransactionResponse])
async def search_transactions(
    request: SearchTransactionRequest,
    db: AuthenticatedDbDep,
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
    current_user: dict = Depends(get_current_user)
):
    """Search transactions with filters"""
    service = POSService(db)
    # TODO: Implement search method
    raise HTTPException(status_code=501, detail="Not implemented")

@router.post("/transactions/batch", response_model=BatchTransactionResponse)
async def create_batch_transactions(
    request: BatchTransactionRequest,
    db: AuthenticatedDbDep,
    current_user: dict = Depends(get_current_user)
):
    """Create multiple transactions in batch"""
    service = POSService(db)
    # TODO: Implement batch creation
    raise HTTPException(status_code=501, detail="Not implemented")

# ============= Shift Endpoints =============

@router.post("/shifts/open", response_model=ShiftResponse)
async def open_shift(
    request: OpenShiftRequest,
    db: AuthenticatedDbDep,
    current_user: dict = Depends(get_current_user)
):
    """Open new POS shift"""
    service = POSService(db)
    return await service.open_shift(request, current_user['id'])

@router.get("/shifts/current", response_model=ShiftResponse)
async def get_current_shift(
    db: AuthenticatedDbDep,
    current_user: dict = Depends(get_current_user)
):
    """Get current open shift for user"""
    service = POSService(db)
    shift = service._get_current_shift(current_user['id'])
    if not shift:
        raise HTTPException(status_code=404, detail="No open shift found")
    return ShiftResponse(**shift)

@router.get("/shifts/{shift_id}", response_model=ShiftResponse)
async def get_shift(
    db: AuthenticatedDbDep,
    shift_id: UUID = Path(...),
    current_user: dict = Depends(get_current_user)
):
    """Get shift details"""
    service = POSService(db)
    shift = service._get_shift(shift_id)
    return ShiftResponse(**shift)

@router.post("/shifts/{shift_id}/close", response_model=ShiftSummaryResponse)
async def close_shift(
    db: AuthenticatedDbDep,
    shift_id: UUID = Path(...),
    request: CloseShiftRequest = ...,
    current_user: dict = Depends(get_current_user)
):
    """Close POS shift with reconciliation"""
    service = POSService(db)
    return await service.close_shift(shift_id, request, current_user['id'])

@router.get("/shifts/{shift_id}/summary", response_model=ShiftSummaryResponse)
async def get_shift_summary(
    db: AuthenticatedDbDep,
    shift_id: UUID = Path(...),
    current_user: dict = Depends(get_current_user)
):
    """Get comprehensive shift summary"""
    service = POSService(db)
    return await service._generate_shift_summary(shift_id)

@router.post("/shifts/{shift_id}/suspend")
async def suspend_shift(
    db: AuthenticatedDbDep,
    shift_id: UUID = Path(...),
    reason: str = Query(...),
    current_user: dict = Depends(get_current_user)
):
    """Temporarily suspend shift"""
    # TODO: Implement shift suspension
    raise HTTPException(status_code=501, detail="Not implemented")

# ============= Receipt Template Endpoints =============

@router.get("/receipt-templates", response_model=List[ReceiptTemplateResponse])
async def get_receipt_templates(
    db: AuthenticatedDbDep,
    is_active: Optional[bool] = Query(True),
    template_type: Optional[str] = Query(None),
    current_user: dict = Depends(get_current_user)
):
    """Get all receipt templates"""
    # TODO: Implement template listing
    raise HTTPException(status_code=501, detail="Not implemented")

@router.post("/receipt-templates", response_model=ReceiptTemplateResponse)
async def create_receipt_template(
    request: ReceiptTemplateCreate,
    db: AuthenticatedDbDep,
    current_user: dict = Depends(get_current_user)
):
    """Create new receipt template"""
    # TODO: Implement template creation
    raise HTTPException(status_code=501, detail="Not implemented")

@router.put("/receipt-templates/{template_id}", response_model=ReceiptTemplateResponse)
async def update_receipt_template(
    db: AuthenticatedDbDep,
    template_id: UUID = Path(...),
    request: ReceiptTemplateUpdate = ...,
    current_user: dict = Depends(get_current_user)
):
    """Update receipt template"""
    # TODO: Implement template update
    raise HTTPException(status_code=501, detail="Not implemented")

@router.delete("/receipt-templates/{template_id}")
async def delete_receipt_template(
    db: AuthenticatedDbDep,
    template_id: UUID = Path(...),
    current_user: dict = Depends(get_current_user)
):
    """Delete receipt template"""
    # TODO: Implement template deletion
    raise HTTPException(status_code=501, detail="Not implemented")

# ============= Category Endpoints =============

@router.get("/categories", response_model=List[POSCategoryResponse])
async def get_categories(
    db: AuthenticatedDbDep,
    is_active: Optional[bool] = Query(True),
    is_featured: Optional[bool] = Query(None),
    parent_id: Optional[UUID] = Query(None),
    current_user: dict = Depends(get_current_user)
):
    """Get POS categories"""
    service = POSService(db)
    return await service.get_categories(is_active, is_featured, parent_id)

@router.post("/categories", response_model=POSCategoryResponse)
async def create_category(
    request: POSCategoryCreate,
    db: AuthenticatedDbDep,
    current_user: dict = Depends(get_current_user)
):
    """Create new POS category"""
    service = POSService(db)
    return await service.create_category(request, current_user['id'])

@router.put("/categories/{category_id}", response_model=POSCategoryResponse)
async def update_category(
    db: AuthenticatedDbDep,
    category_id: UUID = Path(...),
    request: POSCategoryUpdate = ...,
    current_user: dict = Depends(get_current_user)
):
    """Update POS category"""
    # TODO: Implement category update
    raise HTTPException(status_code=501, detail="Not implemented")

# ============= Product Modifier Endpoints =============

@router.get("/products/{product_id}/modifiers", response_model=List[ProductModifierResponse])
async def get_product_modifiers(
    db: AuthenticatedDbDep,
    product_id: UUID = Path(...),
    current_user: dict = Depends(get_current_user)
):
    """Get modifiers for a product"""
    # TODO: Implement modifier listing
    raise HTTPException(status_code=501, detail="Not implemented")

@router.post("/products/{product_id}/modifiers", response_model=ProductModifierResponse)
async def create_product_modifier(
    db: AuthenticatedDbDep,
    product_id: UUID = Path(...),
    request: ProductModifierCreate = ...,
    current_user: dict = Depends(get_current_user)
):
    """Create product modifier"""
    # TODO: Implement modifier creation
    raise HTTPException(status_code=501, detail="Not implemented")

@router.put("/modifiers/{modifier_id}", response_model=ProductModifierResponse)
async def update_product_modifier(
    db: AuthenticatedDbDep,
    modifier_id: UUID = Path(...),
    request: ProductModifierUpdate = ...,
    current_user: dict = Depends(get_current_user)
):
    """Update product modifier"""
    # TODO: Implement modifier update
    raise HTTPException(status_code=501, detail="Not implemented")

# ============= Quick Access Endpoints =============

@router.get("/products/quick-access", response_model=List[QuickProductResponse])
async def get_quick_access_products(
    db: AuthenticatedDbDep,
    category_id: Optional[UUID] = Query(None),
    search: Optional[str] = Query(None),
    is_featured: Optional[bool] = Query(None),
    current_user: dict = Depends(get_current_user)
):
    """Get products for quick access in POS"""
    service = POSService(db)
    # For now, ignore search and is_featured until implemented in service
    return await service.get_quick_products(category_id, is_available=True)

@router.get("/products/barcode/{barcode}", response_model=QuickProductResponse)
async def get_product_by_barcode(
    db: AuthenticatedDbDep,
    barcode: str = Path(...),
    current_user: dict = Depends(get_current_user)
):
    """Get product by barcode scan"""
    # TODO: Implement barcode lookup
    raise HTTPException(status_code=501, detail="Not implemented")

# ============= Report Endpoints =============

@router.get("/reports/daily-summary", response_model=DailySummaryResponse)
async def get_daily_summary(
    db: AuthenticatedDbDep,
    date: date = Query(...),
    terminal_id: Optional[str] = Query(None),
    cashier_id: Optional[UUID] = Query(None),
    current_user: dict = Depends(get_current_user)
):
    """Get daily sales summary"""
    service = POSService(db)
    return await service.get_daily_summary(date, terminal_id, cashier_id)

@router.get("/reports/product-sales", response_model=ProductSalesReport)
async def get_product_sales_report(
    db: AuthenticatedDbDep,
    date_from: date = Query(...),
    date_to: date = Query(...),
    category_id: Optional[UUID] = Query(None),
    product_id: Optional[UUID] = Query(None),
    current_user: dict = Depends(get_current_user)
):
    """Get product sales report"""
    # TODO: Implement product sales report
    raise HTTPException(status_code=501, detail="Not implemented")

@router.get("/reports/hourly-sales")
async def get_hourly_sales(
    db: AuthenticatedDbDep,
    date: date = Query(...),
    terminal_id: Optional[str] = Query(None),
    current_user: dict = Depends(get_current_user)
):
    """Get hourly sales breakdown"""
    # TODO: Implement hourly sales
    raise HTTPException(status_code=501, detail="Not implemented")

@router.get("/reports/payment-methods")
async def get_payment_method_report(
    db: AuthenticatedDbDep,
    date_from: date = Query(...),
    date_to: date = Query(...),
    current_user: dict = Depends(get_current_user)
):
    """Get payment method breakdown"""
    # TODO: Implement payment method report
    raise HTTPException(status_code=501, detail="Not implemented")

# ============= Printer Endpoints =============

@router.get("/printers", response_model=List[PrinterStatusResponse])
async def get_printers(
    db: AuthenticatedDbDep,
    current_user: dict = Depends(get_current_user)
):
    """Get list of available printers"""
    # TODO: Implement printer listing
    raise HTTPException(status_code=501, detail="Not implemented")

@router.get("/printers/{printer_id}/status", response_model=PrinterStatusResponse)
async def get_printer_status(
    db: AuthenticatedDbDep,
    printer_id: str = Path(...),
    current_user: dict = Depends(get_current_user)
):
    """Get printer status"""
    # TODO: Implement printer status check
    raise HTTPException(status_code=501, detail="Not implemented")

@router.post("/printers/{printer_id}/test")
async def test_printer(
    db: AuthenticatedDbDep,
    printer_id: str = Path(...),
    request: PrintTestRequest = ...,
    current_user: dict = Depends(get_current_user)
):
    """Send test print to printer"""
    # TODO: Implement test printing
    raise HTTPException(status_code=501, detail="Not implemented")

# ============= Dashboard Endpoints =============

@router.get("/dashboard/stats")
async def get_pos_dashboard_stats(
    db: AuthenticatedDbDep,
    current_user: dict = Depends(get_current_user)
):
    """Get POS dashboard statistics"""
    # TODO: Implement dashboard stats
    return {
        "today_sales": 0,
        "today_transactions": 0,
        "average_transaction": 0,
        "top_products": [],
        "recent_transactions": []
    }

@router.get("/dashboard/live-feed")
async def get_live_transaction_feed(
    db: AuthenticatedDbDep,
    limit: int = Query(10, ge=1, le=50),
    current_user: dict = Depends(get_current_user)
):
    """Get live transaction feed"""
    # TODO: Implement live feed
    raise HTTPException(status_code=501, detail="Not implemented")

File: frontend/src/lib/api/pos.ts
import { api } from '../api';
import type {
  Transaction,
  CreateTransactionRequest,
  ProcessPaymentRequest,
  Payment,
  QRPaymentInit,
  PaymentStatusCheck,
  Shift,
  ShiftSummary,
  OpenShiftRequest,
  CloseShiftRequest,
  VoidTransactionRequest,
  PrintReceiptRequest,
  Receipt,
  POSCategory,
  ProductModifier,
  QuickProduct,
  ReceiptTemplate,
  DailySummary,
  POSDashboard,
  PrinterStatus
} from '@/types/pos';

export const posApi = {
  // ============= Transactions =============
  
  createTransaction: async (data: CreateTransactionRequest): Promise<Transaction> => {
    const response = await api.post('/pos/transactions/create', data);
    return response.data;
  },

  getTransaction: async (transactionId: string): Promise<Transaction> => {
    const response = await api.get(`/pos/transactions/${transactionId}`);
    return response.data;
  },

  processPayment: async (
    transactionId: string,
    data: ProcessPaymentRequest
  ): Promise<Payment | QRPaymentInit> => {
    const response = await api.post(
      `/pos/transactions/${transactionId}/payment`,
      data
    );
    return response.data;
  },

  checkPaymentStatus: async (transactionId: string): Promise<PaymentStatusCheck> => {
    const response = await api.get(`/pos/transactions/${transactionId}/payment-status`);
    return response.data;
  },

  voidTransaction: async (
    transactionId: string,
    data: VoidTransactionRequest
  ): Promise<Transaction> => {
    const response = await api.post(
      `/pos/transactions/${transactionId}/void`,
      data
    );
    return response.data;
  },

  printReceipt: async (
    transactionId: string,
    data: PrintReceiptRequest
  ): Promise<Receipt> => {
    const response = await api.post(
      `/pos/transactions/${transactionId}/print-receipt`,
      data
    );
    return response.data;
  },

  searchTransactions: async (params: any): Promise<Transaction[]> => {
    const response = await api.post('/pos/transactions/search', params);
    return response.data;
  },

  // ============= Shifts =============

  openShift: async (data: OpenShiftRequest): Promise<Shift> => {
    const response = await api.post('/pos/shifts/open', data);
    return response.data;
  },

  getCurrentShift: async (): Promise<Shift> => {
    const response = await api.get('/pos/shifts/current');
    return response.data;
  },

  getShift: async (shiftId: string): Promise<Shift> => {
    const response = await api.get(`/pos/shifts/${shiftId}`);
    return response.data;
  },

  closeShift: async (shiftId: string, data: CloseShiftRequest): Promise<ShiftSummary> => {
    const response = await api.post(`/pos/shifts/${shiftId}/close`, data);
    return response.data;
  },

  getShiftSummary: async (shiftId: string): Promise<ShiftSummary> => {
    const response = await api.get(`/pos/shifts/${shiftId}/summary`);
    return response.data;
  },

  // ============= Categories & Products =============

  getCategories: async (params?: {
    is_active?: boolean;
    is_featured?: boolean;
    parent_id?: string;
  }): Promise<POSCategory[]> => {
    const response = await api.get('/pos/categories', { params });
    return response.data;
  },

  createCategory: async (data: Partial<POSCategory>): Promise<POSCategory> => {
    const response = await api.post('/pos/categories', data);
    return response.data;
  },

  updateCategory: async (
    categoryId: string,
    data: Partial<POSCategory>
  ): Promise<POSCategory> => {
    const response = await api.put(`/pos/categories/${categoryId}`, data);
    return response.data;
  },

  getProductModifiers: async (productId: string): Promise<ProductModifier[]> => {
    const response = await api.get(`/pos/products/${productId}/modifiers`);
    return response.data;
  },

  getQuickAccessProducts: async (params?: {
    category_id?: string;
    search?: string;
    is_featured?: boolean;
  }): Promise<QuickProduct[]> => {
    const response = await api.get('/pos/products/quick-access', { params });
    return response.data;
  },

  getProductByBarcode: async (barcode: string): Promise<QuickProduct> => {
    const response = await api.get(`/pos/products/barcode/${barcode}`);
    return response.data;
  },

  // ============= Receipt Templates =============

  getReceiptTemplates: async (params?: {
    is_active?: boolean;
    template_type?: string;
  }): Promise<ReceiptTemplate[]> => {
    const response = await api.get('/pos/receipt-templates', { params });
    return response.data;
  },

  createReceiptTemplate: async (data: Partial<ReceiptTemplate>): Promise<ReceiptTemplate> => {
    const response = await api.post('/pos/receipt-templates', data);
    return response.data;
  },

  updateReceiptTemplate: async (
    templateId: string,
    data: Partial<ReceiptTemplate>
  ): Promise<ReceiptTemplate> => {
    const response = await api.put(`/pos/receipt-templates/${templateId}`, data);
    return response.data;
  },

  deleteReceiptTemplate: async (templateId: string): Promise<void> => {
    await api.delete(`/pos/receipt-templates/${templateId}`);
  },

  // ============= Reports =============

  getDailySummary: async (params: {
    date: string;
    terminal_id?: string;
    cashier_id?: string;
  }): Promise<DailySummary> => {
    const response = await api.get('/pos/reports/daily-summary', { params });
    return response.data;
  },

  getProductSalesReport: async (params: {
    date_from: string;
    date_to: string;
    category_id?: string;
    product_id?: string;
  }): Promise<any> => {
    const response = await api.get('/pos/reports/product-sales', { params });
    return response.data;
  },

  getHourlySales: async (params: {
    date: string;
    terminal_id?: string;
  }): Promise<any> => {
    const response = await api.get('/pos/reports/hourly-sales', { params });
    return response.data;
  },

  getPaymentMethodReport: async (params: {
    date_from: string;
    date_to: string;
  }): Promise<any> => {
    const response = await api.get('/pos/reports/payment-methods', { params });
    return response.data;
  },

  // ============= Dashboard =============

  getDashboardStats: async (): Promise<POSDashboard> => {
    const response = await api.get('/pos/dashboard/stats');
    return response.data;
  },

  getLiveTransactionFeed: async (limit?: number): Promise<Transaction[]> => {
    const response = await api.get('/pos/dashboard/live-feed', {
      params: { limit: limit || 10 }
    });
    return response.data;
  },

  // ============= Printers =============

  getPrinters: async (): Promise<PrinterStatus[]> => {
    const response = await api.get('/pos/printers');
    return response.data;
  },

  getPrinterStatus: async (printerId: string): Promise<PrinterStatus> => {
    const response = await api.get(`/pos/printers/${printerId}/status`);
    return response.data;
  },

  testPrinter: async (printerId: string, testType: string = 'alignment'): Promise<void> => {
    await api.post(`/pos/printers/${printerId}/test`, { test_type: testType });
  }
};

// Helper functions for POS operations

export const calculateCartTotals = (items: any[]) => {
  let subtotal = 0;
  
  items.forEach(item => {
    const itemTotal = item.quantity * item.product.price;
    const modifierTotal = item.modifiers.reduce((sum: number, mod: any) => {
      return sum + (mod.price_adjustment || 0) * item.quantity;
    }, 0);
    subtotal += itemTotal + modifierTotal;
  });
  
  const tax = subtotal * 0.1; // 10% VAT
  const total = subtotal + tax;
  
  return {
    subtotal,
    tax,
    total,
    discount: 0,
    service_charge: 0
  };
};

export const formatCurrency = (amount: number): string => {
  return new Intl.NumberFormat('vi-VN', {
    style: 'currency',
    currency: 'VND'
  }).format(amount);
};

export const formatReceiptNumber = (number: string): string => {
  return number || 'PENDING';
};

export const getPaymentMethodLabel = (method: string): string => {
  const labels: { [key: string]: string } = {
    cash: 'Cash',
    bank_transfer: 'Bank Transfer',
    credit_card: 'Credit Card',
    room_charge: 'Room Charge',
    qr_code: 'QR Payment'
  };
  return labels[method] || method;
};

export const getTransactionStatusColor = (status: string): string => {
  const colors: { [key: string]: string } = {
    active: 'blue',
    completed: 'green',
    void: 'red',
    refunded: 'orange'
  };
  return colors[status] || 'gray';
};

export const getPaymentStatusColor = (status: string): string => {
  const colors: { [key: string]: string } = {
    pending: 'yellow',
    partial: 'orange',
    completed: 'green',
    void: 'red'
  };
  return colors[status] || 'gray';
};